media-tv/kodi/kodi-20.3.ebuild

```diff
@@ -0,0 +1,466 @@
+	)
+"
+COMMON_TARGET_DEPEND="${PYTHON_DEPS}
+	>=net-misc/curl-7.68.0[http2]
+	>=sys-libs/zlib-1.2.11
+	dev-db/sqlite:3
+	dev-libs/crossguid
+	>=dev-libs/fribidi-1.0.5
+	>=dev-libs/libcdio-2.1.0:=[cxx]
+	>=dev-libs/libfmt-6.1.2:=
+	dev-libs/libfstrcmp
+	dev-libs/libpcre[cxx]
+	>=dev-libs/openssl-1.1.1k:0=
+	>=dev-libs/spdlog-1.5.0:=
+	dev-libs/tinyxml[stl]
+	media-fonts/roboto
+	media-libs/libglvnd[X?]
+	>=media-libs/freetype-2.10.1
+	media-libs/harfbuzz:=
+	>=media-libs/libass-0.13.4:=
+	media-libs/mesa[egl(+),gbm(+)?,wayland?,X?]
+	>=media-libs/taglib-1.9.0
+	sci-libs/kissfft
+	virtual/libiconv
+	virtual/ttf-fonts
+	x11-libs/libdrm
+	$(python_gen_cond_dep '
+		dev-python/pillow[${PYTHON_USEDEP}]
+		dev-python/pycryptodome[${PYTHON_USEDEP}]
+	')
+	airplay? (
+		>=app-pda/libplist-2.0.0:=
+		net-libs/shairplay
+	)
+	alsa? (
+		>=media-libs/alsa-lib-1.1.4.1
+	)
+	bluetooth? (
+		net-wireless/bluez:=
+	)
+	bluray? (
+		>=media-libs/libbluray-1.1.2:=
+	)
+	caps? (
+		sys-libs/libcap
+	)
+	cec? (
+		>=dev-libs/libcec-4.0[-cubox,raspberry-pi?]
+	)
+	dbus? (
+		sys-apps/dbus
+	)
+	gbm? (
+		>=dev-libs/libinput-1.10.5:=
+		x11-libs/libxkbcommon
+	)
+	gles? (
+		!raspberry-pi? (
+			media-libs/mesa[gles2]
+		)
+	)
+	!gles? (
+		media-libs/glu
+	)
+	lcms? (
+		>=media-libs/lcms-2.10:2
+	)
+	libusb? (
+		virtual/libusb:1
+	)
+	lirc? (
+		app-misc/lirc
+	)
+	mariadb? (
+		dev-db/mariadb-connector-c:=
+	)
+	mysql? (
+		dev-db/mysql-connector-c:=
+	)
+	nfs? (
+		>=net-fs/libnfs-2.0.0:=
+	)
+	pipewire? (
+		>=media-video/pipewire-0.3.24:=
+	)
+	pulseaudio? (
+		media-libs/libpulse
+	)
+	raspberry-pi? (
+		|| (
+			media-libs/raspberrypi-userland
+			media-libs/raspberrypi-userland-bin
+			media-libs/mesa[gles2,video_cards_vc4]
+		)
+	)
+	samba? (
+		>=net-fs/samba-3.4.6[smbclient(+)]
+	)
+	system-ffmpeg? (
+		>=media-video/ffmpeg-${FFMPEG_VERSION}:=[encode,postproc,vaapi?,vdpau?,X?]
+		=media-video/ffmpeg-4*[openssl]
+	)
+	!system-ffmpeg? (
+		app-arch/bzip2
+		media-libs/dav1d:=
+	)
+	udf? (
+		>=dev-libs/libudfread-1.0.0
+	)
+	udev? (
+		virtual/libudev:=
+	)
+	vaapi? (
+		media-libs/libva:=[wayland?,X?]
+	)
+	vdpau? (
+		|| (
+			>=x11-libs/libvdpau-1.1
+			>=x11-drivers/nvidia-drivers-180.51
+		)
+	)
+	wayland? (
+		>=x11-libs/libxkbcommon-0.4.1[wayland]
+	)
+	webserver? (
+		>=net-libs/libmicrohttpd-0.9.55:=[messages(+)]
+	)
+	X? (
+		x11-libs/libX11
+		x11-libs/libXext
+		x11-libs/libXrandr
+	)
+	xslt? (
+		dev-libs/libxslt
+		>=dev-libs/libxml2-2.9.4
+	)
+	zeroconf? (
+		net-dns/avahi[dbus]
+	)
+"
+RDEPEND="
+	${COMMON_DEPEND}
+	${COMMON_TARGET_DEPEND}
+"
+DEPEND="
+	${COMMON_DEPEND}
+	${COMMON_TARGET_DEPEND}
+	dev-libs/rapidjson
+	test? (
+		>=dev-cpp/gtest-1.10.0
+	)
+	wayland? (
+		>=dev-libs/wayland-protocols-1.7
+	)
+	X? (
+		x11-base/xorg-proto
+		x11-libs/libXrender
+	)
+"
+# <dev-lang/swig-4.2.0 https://github.com/xbmc/xbmc/issues/24385
+BDEPEND="
+	${COMMON_DEPEND}
+	<dev-lang/swig-4.2.0
+	dev-util/cmake
+	virtual/pkgconfig
+	virtual/jre
```

Comment:
you can change this to `<=virtual/jre-17:*`. the same goes for the other versions.

Reply:
Yes, but can I enforce it to use lower 21 if someone has multiple versions installed and its set to 21 for the system?

Reply:
> Yes, but can I enforce it to use than lower 21 if someone has multiple versions installed and its set to 21 for the system?

well, idk whether that `BDEPEND` is enforced correctly but this is generally how java related ebuilds work. specifying jdk/jre version in `DEPEND` and `RDEPEND` forces usage of the correct java version even if multiple versions are installed (in case of java packages it also affects source and target version used for compilation). i myself have several versions installed and it works fine. if that does not help to resolve the issue, you can try to move it to `DEPEND` instead.

Reply:
```
# Doesn't build with jdk-21
_JAVA_PKG_WANT_BUILD_VM=( {openjdk{,-jre},icedtea}{,-bin}-{8,11,17} )
JAVA_PKG_WANT_BUILD_VM=${_JAVA_PKG_WANT_BUILD_VM[@]}
# Required to be set, but not used.
JAVA_PKG_WANT_SOURCE="17"
JAVA_PKG_WANT_TARGET="17"
```

This works, but if there are any solutions that are more elegant then I'm all ears.

Reply:
as i wrote before, the standard way is to restrict the version in dependency. that should handle the jdk/jre used. but in all java related projects we don't use `BDEPEND` but `DEPEND` for compilation and `RDEPEND` for runtime. my quick check in the eclasses revealed that `BDEPEND` is not mentioned in java eclasses for this purpose at all.

Reply:
@parona-source the part you're missing is that there's some odd java-config magic where it tries to parse *DEPEND to decide what to do. You are right that in future, we need a proper JAVA_COMPAT-style thing though.